File: 0021/0021_merge_two_sorted_lists.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// MUST BE Ascent sorted list
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil {
		return l2
	} else if l2 == nil {
		return l1
	}

	p1 := l1
	p2 := l2

	var h *ListNode

	if p1.Val <= p2.Val {
		h = p1
		p1 = p1.Next
	} else {
		h = p2
		p2 = p2.Next
	}

	p := h

	for p1 != nil && p2 != nil {
		if p1.Val <= p2.Val {
			p.Next = p1
			p = p.Next
			p1 = p1.Next
		} else {
			p.Next = p2
			p = p.Next
			p2 = p2.Next
		}
	}

	if p1 == nil {
		p.Next = p2
	} else {
		p.Next = p1
	}

	return h
}

File: 0021/0021_merge_two_sorted_lists.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

struct Solution;

impl Solution {
    pub fn merge_two_lists(
        list1: Option<Box<ListNode>>,
        list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        match (list1, list2) {
            (None, None) => None,
            (Some(n), None) | (None, Some(n)) => Some(n),
            (Some(list1), Some(list2)) => {
                if list1.val >= list2.val {
                    Some(Box::new(ListNode{
                        val: list2.val,
                        next: Self::merge_two_lists(Some(list1), list2.next),
                    }))
                } else {
                    Some(Box::new(ListNode{
                        val: list1.val,
                        next: Self::merge_two_lists(list1.next, Some(list2)),
                    }))
                }
            }
        }
    }
}

fn main() {
    Solution::merge_two_lists(None, None);
}
